src/plots/cartesian/axis_autotype.js

```diff
@@ -15,22 +15,42 @@ var Lib = require('../../lib');
 var BADNUM = require('../../constants/numerical').BADNUM;
 
 module.exports = function autoType(array, calendar, opts) {
-    opts = opts || {};
+    var a = array;
+
+    if(Lib.isArrayOrTypedArray(a) && !a.length) return '-';
+    if(!opts.noMultiCategory && multiCategory(a)) return 'multicategory';
+    if(opts.noMultiCategory && Array.isArray(a[0])) {
+        var b = [];
+        for(var i = 0; i < a.length; i++) {
+            if(Array.isArray(a[i])) {
+                for(var j = 0; j < a[i].length; j++) {
+                    b.push(a[i][j]);
+                }
+            }
+        }
+        a = b;
```

Comment:
I guess this is for carpet plots? I still don't see any examples of these with non-numeric axes - either in our tests or in the wild... so I wouldn't spend a lot of time handling that case.

Reply:
Not only `carpet`, but also passing 2d array in `surface` would get there. Something like:
```js
Plotly.newPlot(gd, [{ 
    type: 'surface',
    z: [
        [0, 1],
        [1, 0]
    ],
    y: [0, 1],
    x: [0, 1]
}]);
```.
And we do not want to run `Lib.isDateTime` & `Lib.cleanNumber` on an array.
So I would call that a bug fix.

Reply:
OK - I don't see `noMultiCategory` in the 3D code, so still I don't see how we'd get here, but perhaps we should have that... it would certainly be confusing to put `multicategory` data on a 3D axis!

Reply:
I added tests to lock the correct behaviour in https://github.com/plotly/plotly.js/pull/5240/commits/e02deafeb641c9e651afd6bdd0aa1d7118abccbc.

Reply:
Also there is no need to add `noMultiCategory` flag for non-cartesian traces (e.g. `gl3d`) as the option is set here:
https://github.com/plotly/plotly.js/blob/96d75110706d255ead829ba1d0ba18e8e5a0b67e/src/plots/cartesian/type_defaults.js#L62

Reply:
Ah, the !cartesian test, right. Thanks!